src/dp_vnf.c

```diff
@@ -89,79 +148,78 @@ int dp_del_vnf(const uint8_t ul_addr6[DP_VNF_IPV6_ADDR_SIZE])
+		return ret;
+	}
 
 	ret = rte_hash_del_key(vnf_handle_tbl, ul_addr6);
 	if (DP_FAILED(ret)) {
 		DPGRPC_LOG_WARNING("Cannot remove VNF", DP_LOG_IPV6(ul_addr6), DP_LOG_RET(ret));
 		return ret;
 	}
 
+	rte_free(vnf);
+
 	return DP_OK;
 }
 
-static __rte_always_inline bool dp_vnf_match(const struct dp_vnf *vnf, enum dp_vnf_type type,
-											 uint16_t port_id, uint32_t vni,
-											 struct dp_ip_address *prefix, uint8_t prefix_len)
+static __rte_always_inline bool dp_vnf_match(const struct dp_vnf *vnf, const struct dp_vnf *target_vnf)
 {
-	return (port_id == DP_VNF_MATCH_ALL_PORT_IDS || vnf->port_id == port_id)
-		&& vnf->vni == vni
-		&& vnf->type == type
-		&& vnf->alias_pfx.ol.ip_type == prefix->ip_type
-		&& ((vnf->alias_pfx.ol.ip_type == RTE_ETHER_TYPE_IPV4 && vnf->alias_pfx.ol.ipv4 == prefix->ipv4)
-			|| (vnf->alias_pfx.ol.ip_type == RTE_ETHER_TYPE_IPV6 && rte_rib6_is_equal(vnf->alias_pfx.ol.ipv6, prefix->ipv6)))
-		&& vnf->alias_pfx.length == prefix_len;
+	return (target_vnf->port_id == DP_VNF_MATCH_ALL_PORT_IDS || vnf->port_id == target_vnf->port_id)
+		&& vnf->vni == target_vnf->vni
+		&& vnf->type == target_vnf->type
+		&& vnf->alias_pfx.ol.ip_type == target_vnf->alias_pfx.ol.ip_type
+		&& ((vnf->alias_pfx.ol.ip_type == RTE_ETHER_TYPE_IPV4 && vnf->alias_pfx.ol.ipv4 == target_vnf->alias_pfx.ol.ipv4)
+			|| (vnf->alias_pfx.ol.ip_type == RTE_ETHER_TYPE_IPV6 && rte_rib6_is_equal(vnf->alias_pfx.ol.ipv6, vnf->alias_pfx.ol.ipv6)))
+		&& vnf->alias_pfx.length == target_vnf->alias_pfx.length;
 }
 
 bool dp_vnf_lbprefix_exists(uint16_t port_id, uint32_t vni, struct dp_ip_address *prefix_ip, uint8_t prefix_len)
 {
-	struct dp_vnf *vnf;
-	uint32_t iter = 0;
-	int32_t ret;
-	const void *key;
-
-	while ((ret = rte_hash_iterate(vnf_handle_tbl, &key, (void **)&vnf, &iter)) != -ENOENT) {
-		if (DP_FAILED(ret)) {
-			if (prefix_ip->ip_type == RTE_ETHER_TYPE_IPV4)
-				DPS_LOG_WARNING("Iterating VNF table failed", DP_LOG_RET(ret), DP_LOG_PORTID(port_id),
-								DP_LOG_VNI(vni), DP_LOG_IPV4(prefix_ip->ipv4), DP_LOG_PREFLEN(prefix_len));
-			else
-				DPS_LOG_WARNING("Iterating VNF table failed", DP_LOG_RET(ret), DP_LOG_PORTID(port_id),
-								DP_LOG_VNI(vni), DP_LOG_IPV6(prefix_ip->ipv6), DP_LOG_PREFLEN(prefix_len));
-			return false;
-		}
-		if (dp_vnf_match(vnf, DP_VNF_TYPE_LB_ALIAS_PFX, port_id, vni, prefix_ip, prefix_len))
-			return true;
-	}
-	return false;
+	struct dp_vnf vnf = {
+		.port_id = port_id,
+		.vni = vni,
+		.type = DP_VNF_TYPE_LB_ALIAS_PFX,
+		.alias_pfx.length = prefix_len,
+		.alias_pfx.ol = *prefix_ip,
+	};
+
+	return dp_vnf_value_exists(&vnf);
 }
 
-int dp_del_vnf_by_value(enum dp_vnf_type type, uint16_t port_id, uint32_t vni, struct dp_ip_address *prefix_ip, uint8_t prefix_len)
+int dp_del_vnf_by_value(struct dp_vnf *target_vnf)
 {
 	struct dp_vnf *vnf;
 	uint32_t iter = 0;
 	int32_t ret;
 	const void *key;
 	int delete_count = 0;
 
+	if (DP_FAILED(dp_delete_vnf_value(target_vnf)))
+		return DP_GRPC_ERR_NOT_FOUND;
+
```

Comment:
Could you actually improve this further by storing the VNF key (IPv6) as data inside this new table? That way you can remove the (already marked unsafe) hash table traversal below and directly call `dp_del_vnf()` using that key.

I think you only need to change to `static int dp_add_vnf_value(struct dp_vnf *vnf, const uint8_t *ul_addr6)` I think.

This would also eliminate `dp_vnf_match()` from the looks of things.

Reply:
the code is optimised to use newly added mapping, but as a separate commit.